newsroom/news_api/news/feed/service.py

```diff
@@ -1,137 +1,96 @@
 
-        if search.args.get("exclude_ids"):
-            search.query["bool"]["must_not"].append({"terms": {"_id": search.args["exclude_ids"]}})
+        return values
 
-    def on_fetched(self, doc):
-        self._enhance_hateoas(doc)
-        super().on_fetched(doc)
 
-    def _enhance_hateoas(self, doc):
-        doc.setdefault("_links", {})
-        doc["_links"]["parent"] = {"title": "Home", "href": "/"}
-        # Remove the next and last page references
-        doc["_links"].pop("last", None)
-        doc["_links"].pop("next", None)
+def apply_exclude_ids(request: NewshubSearchRequest[NewsAPIFeedSearchArgs]):
+    if request.args.exclude_ids:
+        request.search.query.must_not.append({"terms": {"_id": request.args.exclude_ids}})
 
-        doc.setdefault("_meta", {})
-        doc["_meta"].pop("page", None)
 
-        self._hateoas_set_item_links(doc)
-        self._hateoas_set_next_page_links(doc)
+default_news_feed_filters = [apply_exclude_ids] + default_search_filters
 
-    def _hateoas_set_item_links(self, doc):
-        for item in doc.get("_items") or []:
-            doc_id = str(item["_id"])
-            item.setdefault("_links", {})
-            item["_links"]["self"] = {
-                "href": "news/item/{}".format(doc_id),
-                "title": "News Item",
-            }
-            item.pop("_updated", None)
-            item.pop("_created", None)
-            item.pop("_etag", None)
 
-    def _hateoas_set_next_page_links(self, doc):
-        args = request.args.to_dict()
+class NewsAPIFeedSearchService(NewsApiSearchServiceAsync):
+    search_args_class = NewsAPIFeedSearchArgs
+    filters = cast(list[SearchFilterFunction], default_news_feed_filters)
+
+    def build_hateoas(self, req: Request, resp: Response, search_req: NewshubSearchRequest[NewsApiSearchRequestArgs]):
+        super().build_hateoas(req, resp, search_req)
+
+        resp.body["_links"].pop("last", None)
+        resp.body["_links"].pop("next", None)
+        resp.body["_meta"].pop("page", None)
+
+        self._hateoas_set_next_page_links(resp, search_req)
+
+    def _hateoas_set_next_page_links(self, resp: Response, search_req: NewshubSearchRequest[NewsApiSearchRequestArgs]):
+        doc = resp.body
+        query_params = search_req.args.to_dict(flatten_lists=True)
 
         if doc["_meta"]["total"] > 0:
             desc_items = list(reversed(doc.get("_items") or []))
-            last_datetime = desc_items[0].get("versioncreated").strftime("%Y-%m-%dT%H:%M:%S")
+            last_datetime = desc_items[0].get("versioncreated")
             exclude_ids = []
 
             for item in desc_items:
-                if item.get("versioncreated").strftime("%Y-%m-%dT%H:%M:%S") != last_datetime:
+                if item.get("versioncreated") != last_datetime:
                     break
 
                 exclude_ids.append(item.get("_id"))
 
-            args["exclude_ids"] = ",".join(exclude_ids)
```

Comment:
`exclude_ids` list is being constructed but not used in the `next_page` HATEOAS link. Any reason for this?

Reply:
what do you mean is not used? It is attached to the URL with the rest of parameters.